app/api/relationships/specs/relationships.spec.js

```diff
@@ -398,6 +399,182 @@ describe('relationships', () => {
           done();
         });
     });
+
+    describe('when saving grouped input', () => {
+      const entity3Expected = {
+        template,
+        type: 'entity',
+        title: 'entity3 title',
+        published: true,
+      };
+
+      const entity4Expected = {
+        template,
+        type: 'entity',
+        title: 'entity4 title',
+        published: true,
+      };
+
+      it('should save grouped input with or without existing hubs', async () => {
+        const result = await relationships.save(
+          [
+            [
+              { entity: 'entity3', hub: hub1 },
+              { entity: 'entity4', hub: hub1 },
+            ],
+            [{ entity: 'entity3' }, { entity: 'entity4' }],
+          ],
+          'en'
+        );
+        expect(result).toMatchObject([
+          {
+            _id: expect.anything(),
+            entity: 'entity3',
+            hub: hub1,
+            entityData: entity3Expected,
+          },
+          {
+            _id: expect.anything(),
+            entity: 'entity4',
+            hub: hub1,
+            entityData: entity4Expected,
+          },
+          {
+            _id: expect.anything(),
+            entity: 'entity3',
+            hub: expect.anything(),
+            entityData: entity3Expected,
+          },
+          {
+            _id: expect.anything(),
+            entity: 'entity4',
+            hub: expect.anything(),
+            entityData: entity4Expected,
+          },
+        ]);
+        expect(result[2].entityData.hub).toEqual(result[3].entityData.hub);
+      });
+
+      it('should allow mixing grouped and non-grouped input', async () => {
+        const result = await relationships.save(
+          [
+            [
+              { entity: 'entity3', hub: hub1 },
+              { entity: 'entity4', hub: hub1 },
+            ],
+            { entity: 'entity3' },
+            { entity: 'entity4' },
+          ],
+          'en'
+        );
+        expect(result).toMatchObject([
+          {
+            _id: expect.anything(),
+            entity: 'entity3',
+            hub: hub1,
+            entityData: entity3Expected,
+          },
+          {
+            _id: expect.anything(),
+            entity: 'entity4',
+            hub: hub1,
+            entityData: entity4Expected,
+          },
+          {
+            _id: expect.anything(),
+            entity: 'entity3',
+            hub: expect.anything(),
+            entityData: entity3Expected,
+          },
+          {
+            _id: expect.anything(),
+            entity: 'entity4',
+            hub: expect.anything(),
+            entityData: entity4Expected,
+          },
+        ]);
+        expect(result[2].entityData.hub).toEqual(result[3].entityData.hub);
```

Comment:
It feels wrong that we assume that connections sent outside a hub (group) are to be within the same hub.
I thought that connections outside a hub should always have a hub because in the bulk they represent an update to a hub (adding a new connection), isn't that true?

Reply:
This assumption only applies when you are creating a new hub. So, what we assume that in each subgroup:
- all elements have a hub or all elements do not have a hub
- if they all have a hub, they can belong to any hub, even mixed - I have changed the first test to show this option
- if they all do not have a hub, we assume that they are to belong to a new hub - you see this check in the test

Technically, you could mix things and say that the "non-hubbed" relationships will still go to the same new hub, but when we did the main relationship save optimization, we have not found any use case for than in Uwazi, and we have opted for the minimal solution that covers the existing cases, with an error message for safety.

Reply:
Forgot to say it LGTM :D Thanks for the explanation